File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/CalculateDetailsActivity.java
package openfoodfacts.github.scrachx.openfood.features.product.view;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.databinding.CalculateDetailsBinding;
import openfoodfacts.github.scrachx.openfood.features.adapters.CalculatedNutrimentsGridAdapter;
import openfoodfacts.github.scrachx.openfood.features.shared.BaseActivity;
import openfoodfacts.github.scrachx.openfood.models.HeaderNutrimentListItem;
import openfoodfacts.github.scrachx.openfood.models.NutrimentListItem;
import openfoodfacts.github.scrachx.openfood.models.Nutriments;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.models.Units;
import openfoodfacts.github.scrachx.openfood.utils.Modifier;
import openfoodfacts.github.scrachx.openfood.utils.ProductUtils;
import openfoodfacts.github.scrachx.openfood.utils.UnitUtils;

import static androidx.recyclerview.widget.DividerItemDecoration.VERTICAL;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.CARBOHYDRATES;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.CARBO_MAP;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.ENERGY_KCAL;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.ENERGY_KJ;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.FAT;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.FAT_MAP;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.MINERALS_MAP;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.PROTEINS;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.PROT_MAP;
import static openfoodfacts.github.scrachx.openfood.models.Nutriments.VITAMINS_MAP;

public class CalculateDetailsActivity extends BaseActivity {
    private static final String KEY_SAMPLE_OBJECT = "sampleObject";
    private static final String KEY_SPINNER_VALUE = "spinnervalue";
    private static final String KEY_WEIGHT = "weight";
    Nutriments nutriments;
    private List<NutrimentListItem> nutrimentListItems;
    private Product product;
    private String spinnerValue;
    private float weight;

    /**
     * @deprecated use {@link #start(Context, Product, String, float)}
     */
    @Deprecated
    public static void start(@NonNull Context context, @NonNull Product product, @NonNull String spinnerValue, @NonNull String weight) {
        start(context, product, spinnerValue, Float.parseFloat(weight));
    }

    public static void start(@NonNull Context context, @NonNull Product product, @NonNull String spinnerValue, float weight) {
        Intent starter = new Intent(context, CalculateDetailsActivity.class);
        starter.putExtra(KEY_SAMPLE_OBJECT, product);
        starter.putExtra(KEY_SPINNER_VALUE, spinnerValue);
        starter.putExtra(KEY_WEIGHT, weight);
        context.startActivity(starter);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getResources().getBoolean(R.bool.portrait_only)) {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        openfoodfacts.github.scrachx.openfood.databinding.CalculateDetailsBinding binding = CalculateDetailsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setTitle(getString(R.string.app_name_long));

        setSupportActionBar(binding.toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        final Intent intent = getIntent();

        product = (Product) intent.getSerializableExtra(KEY_SAMPLE_OBJECT);
        spinnerValue = intent.getStringExtra(KEY_SPINNER_VALUE);
        weight = intent.getFloatExtra(KEY_WEIGHT, -1);
        if (product == null || spinnerValue == null || weight == -1) {
            Log.e(CalculateDetailsActivity.class.getSimpleName(), "fragment instantiated with wrong intent extras");
            finish();
        }
        nutriments = product.getNutriments();
        nutrimentListItems = new ArrayList<>();
        binding.resultTextView.setText(getString(R.string.display_fact, weight + " " + spinnerValue));
        binding.nutrimentsRecyclerViewCalc.setHasFixedSize(true);

        // use a linear layout manager
        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getApplicationContext());
        binding.nutrimentsRecyclerViewCalc.setLayoutManager(mLayoutManager);

        binding.nutrimentsRecyclerViewCalc.setNestedScrollingEnabled(false);

        // use VERTICAL divider
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(binding.nutrimentsRecyclerViewCalc.getContext(), VERTICAL);
        binding.nutrimentsRecyclerViewCalc.addItemDecoration(dividerItemDecoration);

        // Header hack
        nutrimentListItems.add(new NutrimentListItem(ProductUtils.isPerServingInLiter(product)));

        // Energy
        Nutriments.Nutriment energyKcal = nutriments.get(ENERGY_KCAL);
        if (energyKcal != null) {
            nutrimentListItems.add(
                new NutrimentListItem(getString(R.string.nutrition_energy_short_name),
                    calculateCalories(weight, spinnerValue),
                    energyKcal.getForServingInUnits(),
                    Units.ENERGY_KCAL,
                    nutriments.getModifierIfNotDefault(ENERGY_KCAL)));
        }
        Nutriments.Nutriment energyKj = nutriments.get(ENERGY_KJ);
        if (energyKj != null) {
            nutrimentListItems.add(
                new NutrimentListItem(getString(R.string.nutrition_energy_short_name),
                    calculateKj(weight, spinnerValue),
                    energyKj.getForServingInUnits(),
                    Units.ENERGY_KJ.toLowerCase(),
                    nutriments.getModifierIfNotDefault(ENERGY_KJ)));
        }

        // Fat
        Nutriments.Nutriment fat = nutriments.get(FAT);
        if (fat != null) {
            String modifier = nutriments.getModifier(FAT);
            nutrimentListItems.add(new HeaderNutrimentListItem(getString(R.string.nutrition_fat),
                fat.getForAnyValue(weight, spinnerValue),
                fat.getForServingInUnits(),
                fat.getUnit(),
                Modifier.DEFAULT_MODIFIER.equals(modifier) ? "" : modifier));

            nutrimentListItems.addAll(getNutrimentItems(nutriments, FAT_MAP));
        }

        // Carbohydrates
        Nutriments.Nutriment carbohydrates = nutriments.get(CARBOHYDRATES);
        if (carbohydrates != null) {
            String modifier = nutriments.getModifier(CARBOHYDRATES);
            nutrimentListItems.add(new HeaderNutrimentListItem(getString(R.string.nutrition_carbohydrate),
                carbohydrates.getForAnyValue(weight, spinnerValue),
                carbohydrates.getForServingInUnits(),
                carbohydrates.getUnit(),
                Modifier.DEFAULT_MODIFIER.equals(modifier) ? "" : modifier));

            nutrimentListItems.addAll(getNutrimentItems(nutriments, CARBO_MAP));
        }

        // fiber
        nutrimentListItems.addAll(getNutrimentItems(nutriments, Collections.singletonMap(Nutriments.FIBER, R.string.nutrition_fiber)));

        // Proteins
        Nutriments.Nutriment proteins = nutriments.get(PROTEINS);
        if (proteins != null) {
            String modifier = nutriments.getModifier(PROTEINS);
            nutrimentListItems.add(
                new HeaderNutrimentListItem(getString(R.string.nutrition_proteins),
                    proteins.getForAnyValue(weight, spinnerValue),
                    proteins.getForServingInUnits(),
                    proteins.getUnit(),
                    Modifier.DEFAULT_MODIFIER.equals(modifier) ? "" : modifier));

            nutrimentListItems.addAll(getNutrimentItems(nutriments, PROT_MAP));
        }

        // salt and alcohol
        Map<String, Integer> map = new HashMap<>();
        map.put(Nutriments.SALT, R.string.nutrition_salt);
        map.put(Nutriments.SODIUM, R.string.nutrition_sodium);
        map.put(Nutriments.ALCOHOL, R.string.nutrition_alcohol);
        nutrimentListItems.addAll(getNutrimentItems(nutriments, map));

        // Vitamins
        if (nutriments.hasVitamins()) {
            nutrimentListItems.add(new HeaderNutrimentListItem(getString(R.string.nutrition_vitamins)));

            nutrimentListItems.addAll(getNutrimentItems(nutriments, VITAMINS_MAP));
        }

        // Minerals
        if (nutriments.hasMinerals()) {
            nutrimentListItems.add(new HeaderNutrimentListItem(getString(R.string.nutrition_minerals)));

            nutrimentListItems.addAll(getNutrimentItems(nutriments, MINERALS_MAP));
        }
        binding.nutrimentsRecyclerViewCalc.setAdapter(new CalculatedNutrimentsGridAdapter(nutrimentListItems));
    }

    private List<NutrimentListItem> getNutrimentItems(Nutriments nutriments, Map<String, Integer> nutrimentMap) {
        List<NutrimentListItem> items = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : nutrimentMap.entrySet()) {
            Nutriments.Nutriment nutriment = nutriments.get(entry.getKey());
            if (nutriment != null) {
                final String modifier = nutriments.getModifier(entry.getKey());
                items.add(new NutrimentListItem(getString(entry.getValue()),
                    nutriment.getForAnyValue(weight, spinnerValue),
                    nutriment.getForServingInUnits(),
                    nutriment.getUnit(),
                    Modifier.DEFAULT_MODIFIER.equals(modifier) ? "" : modifier));
            }
        }

        return items;
    }

    private String calculateCalories(float weight, String unit) {
        float caloriePer100g = Float.parseFloat(product.getNutriments().get(ENERGY_KCAL).getFor100gInUnits());
        float weightInG = UnitUtils.convertToGrams(weight, unit);
        return Float.toString(((caloriePer100g / 100) * weightInG));
    }

    private String calculateKj(float weight, String unit) {
        float caloriePer100g = Float.parseFloat(product.getNutriments().get(ENERGY_KJ).getFor100gInUnits());
        float weightInG = UnitUtils.convertToGrams(weight, unit);
        return Float.toString(((caloriePer100g / 100) * weightInG));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Respond to the action bar's Up/Home button
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/CalculateDetailsActivity.kt
package openfoodfacts.github.scrachx.openfood.features.product.view

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.DividerItemDecoration.VERTICAL
import androidx.recyclerview.widget.LinearLayoutManager
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.CalculateDetailsBinding
import openfoodfacts.github.scrachx.openfood.features.adapters.CalculatedNutrimentsGridAdapter
import openfoodfacts.github.scrachx.openfood.features.shared.BaseActivity
import openfoodfacts.github.scrachx.openfood.models.*
import openfoodfacts.github.scrachx.openfood.models.MeasurementUnit
import openfoodfacts.github.scrachx.openfood.utils.Measurement
import openfoodfacts.github.scrachx.openfood.utils.grams
import openfoodfacts.github.scrachx.openfood.utils.isPerServingInLiter
import openfoodfacts.github.scrachx.openfood.utils.measure
import java.util.*
import kotlin.properties.Delegates

class CalculateDetailsActivity : BaseActivity() {
    private val nutrimentListItems = mutableListOf<NutrimentListItem>()

    private lateinit var nutriments: ProductNutriments
    private lateinit var product: Product

    private var weight by Delegates.notNull<Float>()
    private lateinit var unitOfMeasurement: MeasurementUnit

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val binding = CalculateDetailsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        title = getString(R.string.app_name_long)

        setSupportActionBar(binding.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        val product = intent.getSerializableExtra(KEY_PRODUCT) as Product?

        val weight = intent.getFloatExtra(KEY_WEIGHT, -1f)
        val unit = intent.getStringExtra(KEY_SPINNER_VALUE)?.let { MeasurementUnit.findBySymbol(it) }

        requireNotNull(product) { "${this::class.simpleName} created without product intent extra." }
        requireNotNull(unit) { "${this::class.simpleName} created without spinner value intent extra." }
        require(weight != -1f) { "${this::class.simpleName} created with weight = -1" }

        this.product = product
        this.unitOfMeasurement = unit
        this.weight = weight
        this.nutriments = product.nutriments

        binding.resultTextView.text = getString(R.string.display_fact, "$weight ${unit.sym}")

        binding.nutriments.setHasFixedSize(true)
        binding.nutriments.layoutManager = LinearLayoutManager(this)
        binding.nutriments.isNestedScrollingEnabled = false

        // use VERTICAL divider
        binding.nutriments.addItemDecoration(DividerItemDecoration(this, VERTICAL))

        // Header hack
        nutrimentListItems += NutrimentListItem(product.isPerServingInLiter() ?: false)

        val portion = measure(weight, unit)

        // Energy
        val energyKcal = nutriments[Nutriment.ENERGY_KCAL]
        if (energyKcal != null) {
            nutrimentListItems += NutrimentListItem(
                getString(R.string.nutrition_energy_short_name),
                calculateCalories(portion).value,
                energyKcal.perServingInUnit?.value,
                MeasurementUnit.ENERGY_KCAL,
                energyKcal.modifier,
            )
        }
        val energyKj = nutriments[Nutriment.ENERGY_KJ]
        if (energyKj != null) {
            nutrimentListItems += NutrimentListItem(
                getString(R.string.nutrition_energy_short_name),
                calculateKj(portion).value,
                energyKj.perServingInUnit?.value,
                MeasurementUnit.ENERGY_KJ,
                energyKj.modifier,
            )
        }

        // Fat
        val fat = nutriments[Nutriment.FAT]
        if (fat != null) {
            nutrimentListItems += BoldNutrimentListItem(
                getString(R.string.nutrition_fat),
                fat.getForPortion(portion).value,
                fat.perServingInUnit?.value,
                fat.unit,
                fat.modifier
            )
            nutrimentListItems += getNutrimentItems(nutriments, FAT_MAP)
        }

        // Carbohydrates
        val carbohydrates = nutriments[Nutriment.CARBOHYDRATES]
        if (carbohydrates != null) {
            nutrimentListItems += BoldNutrimentListItem(
                getString(R.string.nutrition_carbohydrate),
                carbohydrates.getForPortion(portion).value,
                carbohydrates.perServingInUnit?.value,
                carbohydrates.unit,
                carbohydrates.modifier
            )
            nutrimentListItems += getNutrimentItems(nutriments, CARBO_MAP)
        }

        // Fiber
        nutrimentListItems += getNutrimentItems(nutriments, mapOf(Nutriment.FIBER to R.string.nutrition_fiber))

        // Proteins
        val proteins = nutriments[Nutriment.PROTEINS]
        if (proteins != null) {
            nutrimentListItems += BoldNutrimentListItem(
                getString(R.string.nutrition_proteins),
                proteins.getForPortion(portion).value,
                proteins.perServingInUnit?.value,
                proteins.unit,
                proteins.modifier
            )
            nutrimentListItems += getNutrimentItems(nutriments, PROT_MAP)
        }

        // salt and alcohol
        nutrimentListItems += getNutrimentItems(
            nutriments, mapOf(
                Nutriment.SALT to R.string.nutrition_salt,
                Nutriment.SODIUM to R.string.nutrition_sodium,
                Nutriment.ALCOHOL to R.string.nutrition_alcohol
            )
        )

        // Vitamins
        if (nutriments.hasVitamins) {
            nutrimentListItems += BoldNutrimentListItem(getString(R.string.nutrition_vitamins))
            nutrimentListItems += getNutrimentItems(nutriments, VITAMINS_MAP)
        }

        // Minerals
        if (nutriments.hasMinerals) {
            nutrimentListItems += BoldNutrimentListItem(getString(R.string.nutrition_minerals))
            nutrimentListItems += getNutrimentItems(nutriments, MINERALS_MAP)
        }
        binding.nutriments.adapter = CalculatedNutrimentsGridAdapter(nutrimentListItems)
    }

    private fun getNutrimentItems(nutriments: ProductNutriments, nutrimentMap: Map<Nutriment, Int>): List<NutrimentListItem> {
        return nutrimentMap.mapNotNull { (name, stringRes) ->
            val nutriment = nutriments[name] ?: return@mapNotNull null

            NutrimentListItem(
                getString(stringRes),
                nutriment.getForPortion(measure(weight, unitOfMeasurement)).value,
                nutriment.perServingInUnit?.value,
                nutriment.unit,
                nutriment.modifier
            )
        }
    }

    private fun calculateCalories(portion: Measurement): Measurement {
        val energy100gCal = product.nutriments[Nutriment.ENERGY_KCAL]!!.per100gInG
        val portionGrams = portion.grams.value
        return Measurement(energy100gCal.value / 100 * portionGrams, energy100gCal.unit)
    }

    private fun calculateKj(portion: Measurement): Measurement {
        val energy100gKj = product.nutriments[Nutriment.ENERGY_KJ]!!.per100gInG
        val weightGrams = portion.grams.value
        return Measurement(energy100gKj.value / 100 * weightGrams, energy100gKj.unit)
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        // Respond to the action bar's Up/Home button
        android.R.id.home -> {
            finish()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    companion object {
        private const val KEY_PRODUCT = "product"
        private const val KEY_SPINNER_VALUE = "spinnerValue"
        private const val KEY_WEIGHT = "weight"

        fun start(context: Context, product: Product, spinnerValue: String, weight: Float) {
            context.startActivity(Intent(context, CalculateDetailsActivity::class.java).apply {
                putExtra(KEY_PRODUCT, product)
                putExtra(KEY_SPINNER_VALUE, spinnerValue)
                putExtra(KEY_WEIGHT, weight)
            })
        }
    }
}